wp/Makefile

```diff
@@ -1,25 +1,63 @@
-BAP_WP = lib/bap_wp
-WP = plugin
-SAMPLE_BINS = resources/sample_binaries
+SHELL=/bin/bash
+LIB := lib/bap_wp
+PLUGIN := plugin
+SAMPLE_BINS := resources/sample_binaries
+FMT_STR := "\u001b[31m%s\u001b[0m\n"
 
-.PHONY: install doc test clean reinstall
+.PHONY: test.lib test.plugin test all clean
 
-install:
-	$(MAKE) -C $(BAP_WP) $@.local
-	$(MAKE) -C $(WP) $@
+all: install
+
+install: uninstall.plugin install.lib build.plugin install.plugin
+
+uninstall.lib:
+	@printf $(FMT_STR) "UNINSTALLING LIB"
+	$(MAKE) -C $(LIB) uninstall
+
+uninstall.plugin:
+	@printf $(FMT_STR) "UNINSTALLING PLUGIN"
+	$(MAKE) -C $(PLUGIN) uninstall
+
+
+build: build.lib build.plugin
+
+build.lib:
+	@printf $(FMT_STR) "BUILDING LIB"
+	$(MAKE) -C $(LIB) $*
+
+build.plugin: install.lib
+	@printf $(FMT_STR) "BUILDING PLUGIN"
+	$(MAKE) -C $(PLUGIN) build
 
 doc:
-	$(MAKE) -C $(BAP_WP) $@
+	$(MAKE) -C $(LIB) $@
+
+test: install test.lib test.plugin
 
-test: install
-	$(MAKE) -C $(BAP_WP) $@
-	$(MAKE) -C $(WP) $@
+test.lib:
+	@printf $(FMT_STR) "TESTING LIB"
+	$(MAKE) -C $(LIB) test
 
-test.clean:
-	$(MAKE) -C $(SAMPLE_BINS) clean
+test.plugin:
+	@printf $(FMT_STR) "TESTING PLUGIN"
+	$(MAKE) -C $(PLUGIN) test
 
 clean:
-	$(MAKE) -C $(WP) $@
-	$(MAKE) -C $(BAP_WP) $@.local
+	@printf $(FMT_STR) "CLEANING LIB"
+	$(MAKE) -C $(LIB) clean
+	@printf $(FMT_STR) "CLEANING PLUGIN"
+	$(MAKE) -C $(PLUGIN) clean
 
 reinstall: clean install
```

Comment:
It's odd to me that the behavior of `reinstall` is different than doing `uninstall` and then `install`.  In particular `reinstall` will clean, but the `uninstall` target does not.  Maybe this should be:

```
reinstall: uninstall.plugin uninstall.lib install
```

And you should have to call `clean` if your really want to rebuild, to match the behavior of the other `install`/`uninstall` targets?

Reply:
To make things more consistent, I've rearranged such that the behavior of reinstall matches between makefiles, and is equivalent to `uninstall install`.